spatial/r2/triangle.go

```diff
@@ -0,0 +1,64 @@
+// Copyright ©2022 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package r2
+
+import "math"
+
+// Triangle represents a triangle in 2D space and
+// is composed by 3 vectors corresponding to the position
+// of each of the vertices.
+type Triangle [3]Vec
+
+// Degenerate returns true if triangle's vertices are collinear to within
+// a certain tolerance.
+func (t Triangle) Degenerate(tol float64) bool {
+	// TODO(soypat): Must be a better way to do this.
+	// https://stackoverflow.com/questions/33037449/given-three-side-of-a-triangle-how-can-i-define-whether-it-is-a-degenerate-tria
```

Comment:
How does this compare to finding the cross product of two vertices relative to the third, finding the area or comparing the slope of two pairs? It would be nice to have a measure where the tolerance is intuitively understandable.

Reply:
I think checking the area lends itself to user error, as a long, needle-like triangle could have a larger area than a well-formed small triangle. I would vouch for `tol` having some spatial tie, i.e. triangle is degenerate if a cylinder of radius `tol` is able to contain all it's vertices. This would then be easy to test and intuitively understand- and I think I have some idea on how to program it. What do you think?

Reply:
> triangle is degenerate if a cylinder of radius `tol` is able to contain all of its vertices

I don't think this is right; an arbitrarily long major side with two minors that sum to |major|+eps  is degenerate to approximately eps.

A either of the longest sides or the shortest sides would give you something where that could be reasonably safely normalised with the mean of the sides' lengths (choose longest or shortest based on whether the triangle is a needle or a pancake).

Reply:
> A either of the longest sides or the shortest sides would give you something where that could be reasonably safely normalised with the mean of the sides' lengths (choose longest or shortest based on whether the triangle is a needle or a pancake).

Also I'm not sure I understand

Reply:
I'm just trying to come up with a reasonable metric for degeneracy. Another alternative is to find the distance between the longest edge and its opposite vertex. This has a very clear intuitive meaning.

Reply:
I like that metric. I've included it in the r3 PR and works as expected.

https://github.com/gonum/gonum/pull/1792/files#diff-8e1c51bb20ff0be4cc795a9701e444f915453f31dcfba6c4bd8ea348a1ebe71aR27